src/librustc/ty/context.rs

```diff
@@ -1432,6 +1432,16 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
         self.queries.on_disk_cache.serialize(self.global_tcx(), encoder)
     }
 
+    /// Either immediate signals the given diagnostic, or puts it on the
+    /// given `buffer` for later signalling in the future.
+    pub fn buffer(self, mut diag: DiagnosticBuilder, buffer: &mut Vec<Diagnostic>) {
+        if self.sess.opts.debugging_opts.dont_buffer_diagnostics {
```

Comment:
I believe this should check `treat-err-as-bug` and abort. And/or you could probably make this:

```rust
if self.sess.opts.debugging_opts.treat_err_as_bug || self.sess.opts.debugging_opts.dont_buffer_diagnostics {
```

I've applied this patch myself numerous times.

Reply:
Oh, wait, no I haven't. I think this function is in the wrong place. This code can live on `buffer` -- it has access to the [`Handler`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.Handler.html), which has the field `flags` (of type [`HandlerFlags`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.HandlerFlags.html)) that contain flags relevant to it. We can just add this flag in there.

Reply:
Ok will do (probably on Monday)